File: src/data/sync/tombstones.ts
/**
 * Lapidas de borrado.
 *
 * Sin esto la sincronizacion resucita lo borrado: si borras un gasto en el
 * telefono, el portatil todavia lo tiene, y en el siguiente "subir" lo
 * vuelve a empujar a la nube. Un borrado tiene que viajar igual que una
 * edicion, y para viajar tiene que existir como dato.
 *
 * Se guardan por (entidad, id) y se limpian cuando ya se aplicaron en
 * ambos lados.
 */
export type DeletableEntity = 'transactions' | 'categories' | 'paymentMethods' | 'recurringRules';

export interface Tombstone {
  /** `${entity}:${entityId}` — llave primaria, idempotente. */
  id: string;
  entity: DeletableEntity;
  entityId: string;
  deletedAt: string;
}

export function tombstoneId(entity: DeletableEntity, entityId: string): string {
  return `${entity}:${entityId}`;
}

export function makeTombstone(entity: DeletableEntity, entityId: string, deletedAt: string): Tombstone {
  return { id: tombstoneId(entity, entityId), entity, entityId, deletedAt };
}

/** Los ids borrados de una entidad, para filtrar lo que llega de la nube. */
export function deletedIdsOf(tombstones: Tombstone[], entity: DeletableEntity): Set<string> {
  const out = new Set<string>();
  for (const t of tombstones) if (t.entity === entity) out.add(t.entityId);
  return out;
}

/**
 * Une las lapidas locales con las remotas, quedandose con la fecha mas
 * vieja de cada una (la del borrado original, no la de la copia).
 */
export function mergeTombstones(a: Tombstone[], b: Tombstone[]): Tombstone[] {
  const byId = new Map<string, Tombstone>();
  for (const t of [...a, ...b]) {
    const prev = byId.get(t.id);
    if (!prev || t.deletedAt < prev.deletedAt) byId.set(t.id, t);
  }
  return Array.from(byId.values());
}

File: mobile/lib/data/tombstones.dart
/// Lápidas de borrado.
///
/// Sin esto la sincronización resucita lo borrado: si borras un gasto en el
/// teléfono, el portátil todavía lo tiene, y en el siguiente "subir" lo
/// vuelve a empujar a la nube. Un borrado tiene que viajar igual que una
/// edición, y para viajar tiene que existir como dato.
///
/// Espejo de src/data/sync/tombstones.ts.
library;

enum DeletableEntity {
  transactions,
  categories,
  paymentMethods,
  recurringRules;

  static DeletableEntity desdeJson(String v) =>
      DeletableEntity.values.firstWhere((e) => e.name == v);
  String get json => name;
}

class Tombstone {
  const Tombstone({required this.entity, required this.entityId, required this.deletedAt});

  final DeletableEntity entity;
  final String entityId;
  final String deletedAt;

  /// `entidad:id` — llave primaria, idempotente.
  String get id => tombstoneId(entity, entityId);

  Map<String, Object?> toJson() => {
        'id': id,
        'entity': entity.json,
        'entityId': entityId,
        'deletedAt': deletedAt,
      };

  static Tombstone fromJson(Map<String, Object?> j) => Tombstone(
        entity: DeletableEntity.desdeJson(j['entity']! as String),
        entityId: j['entityId']! as String,
        deletedAt: j['deletedAt']! as String,
      );

  @override
  bool operator ==(Object other) =>
      other is Tombstone && other.id == id && other.deletedAt == deletedAt;

  @override
  int get hashCode => Object.hash(id, deletedAt);
}

String tombstoneId(DeletableEntity entity, String entityId) => '${entity.name}:$entityId';

/// Los ids borrados de una entidad, para filtrar lo que llega de la nube.
Set<String> deletedIdsOf(List<Tombstone> tombstones, DeletableEntity entity) =>
    tombstones.where((t) => t.entity == entity).map((t) => t.entityId).toSet();

/// Une las lápidas locales con las remotas, quedándose con la fecha más
/// vieja de cada una (la del borrado original, no la de la copia).
List<Tombstone> mergeTombstones(List<Tombstone> a, List<Tombstone> b) {
  final byId = <String, Tombstone>{};
  for (final t in [...a, ...b]) {
    final prev = byId[t.id];
    if (prev == null || t.deletedAt.compareTo(prev.deletedAt) < 0) byId[t.id] = t;
  }
  return byId.values.toList();
}
